rustls/src/crypto/mod.rs

```diff
@@ -505,6 +505,83 @@ pub trait ActiveKeyExchange: Send + Sync {
+    /// like to avoid those wastes if possible.
+    ///
+    /// It is early days for post-quantum-secure hybrid key exchange deployment.
+    /// This means (commonly) continuing to offer both the hybrid and classical
+    /// key exchanges, so the handshake can be completed without a `HelloRetryRequest`
+    /// for servers that support the offered hybrid or classical schemes.
+    ///
+    /// Implementing `hybrid_pair()` enables two optimizations:
+    ///
+    /// 1. Sending both the hybrid and classical key shares in the `ClientHello`.
+    ///
+    /// 2. Performing the classical key exchange setup only once.  This is important
+    ///    because the classical key exchange setup is relatively expensive.
+    ///    This optimization is permitted and described in
+    ///    <https://www.ietf.org/archive/id/draft-ietf-tls-hybrid-design-10.html#section-3.2>
+    ///
+    /// Both of these only happen if the classical algorithm appears separately in
+    /// the client's [`CryptoProvider::kx_groups`], and if the hybrid algorithm appears
+    /// first in that list.
+    ///
+    /// # How it works
+    /// This function is only called by rustls for clients.  It is called when
+    /// constructing the initial `ClientHello`.  rustls follows these steps:
+    ///
+    /// 1. If the return value is None, nothing further happens.
+    /// 2. If the given [`ActiveKeyExchange::group()`] does not appear in
+    ///    [`CryptoProvider::kx_groups`], nothing further happens.
+    /// 3. The given [`ActiveKeyExchange::pub_key()`] is added to the `ClientHello`,
+    ///    after the hybrid entry.
+    ///
+    /// Then, one of three things may happen when the server replies to the `ClientHello`:
+    ///
+    /// 1. The server sends a `HelloRetryRequest`.  Everything is thrown away and
+    ///    we start again.  Note that `into_hybrid_pair()` is never called in the case
+    ///    of a `HelloRetryRequest`.
+    /// 2. The server agrees to our hybrid key exchange: rustls calls
+    ///    [`ActiveKeyExchange::complete()`] consuming `self`.  `into_hybrid_pair()`
+    ///    is not called.
+    /// 3. The server agrees to our classical key exchange: rustls calls
+    ///    [`ActiveKeyExchange::into_hybrid_pair()`] which discards the hybrid
+    ///    key data, and yields just the classical key exchange which is then
+    ///    consumed by [`ActiveKeyExchange::complete()`].
+    fn hybrid_pair(&self) -> Option<&dyn ActiveKeyExchange> {
+        None
+    }
+
+    /// Consumes this hybrid key exchange, yielding the underlying classical
+    /// key exchange.
+    ///
+    /// If [`ActiveKeyExchange::hybrid_pair()`] returns Some(_), this
+    /// _must_ return Some(_), and return the same underlying `dyn ActiveKeyExchange`
+    /// object.
+    ///
+    /// See [`ActiveKeyExchange::hybrid_pair()`] for what this is for.
+    fn into_hybrid_pair(self: Box<Self>) -> Option<Box<dyn ActiveKeyExchange>> {
```

Comment:
Given the description of "consumes this hybrid exhcnage, yielding the underlying classical exchange", the name "into_hybrid_pair()` seems pretty confusing to me.

Reply:
mmm, it's not the clearest. i think the important points are: "if self is a hybrid key exchange, give me half that might be supported separately". so, maybe `into_hybrid_half`?

Reply:
Would describing it as a "fallback_exchange()" be confusing? :house: :paintbrush:

Reply:
for me "fallback" isn't conveying the right thing, but also now i think "hybrid" is a property of self, not the returned thing. will re-read the RFC draft and see if anything jumps out at me.

Reply:
trying out `hybrid_component()` and `split_hybrid_component()` for size

Reply:
Going to resolve this thread; I'm happy with this naming and I think it's matching the language in https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design-11 is a justifiable approach.